File: bench/workbank/tools/compare.py
#!/usr/bin/env python3
"""compare.py — compare two workbank runs or configs case by case.

Usage:
  compare.py <run_dir_A> <run_dir_B>      two run directories (summary.json+run.json)
  compare.py <config_name_A> <config_name_B>   two config names in ledger/cases.jsonl

For each argument a per-case pass value is built: 1.0/0.0 for run dirs, the
mean over k replicas for config names. Cases are aligned by case_id and the
report contains:
  - the flip list (A passed & B failed, A failed & B passed) with trace refs
  - a bootstrap 95% CI of the pass-rate difference (A - B), resampling whole
    families (same-family variants are highly correlated; cases without a
    family each form their own group), 2000 replicates, seed 0.

Output: a JSON object followed by a human-readable summary. Stdlib only.
"""
import argparse
import json
import random
import sys
from pathlib import Path

TOOLS_DIR = Path(__file__).resolve().parent
CASES_JSONL = TOOLS_DIR.parent / "ledger" / "cases.jsonl"
N_BOOT = 2000
BOOT_SEED = 0
PASS_THRESHOLD = 0.5


def read_jsonl(path):
    path = Path(path)
    if not path.is_file():
        return []
    rows = []
    for i, line in enumerate(path.read_text(encoding="utf-8").splitlines(), 1):
        line = line.strip()
        if not line:
            continue
        try:
            rows.append(json.loads(line))
        except json.JSONDecodeError:
            print("warning: %s:%d is not valid JSON, skipped" % (path, i), file=sys.stderr)
    return rows


def load_run_dir(run_dir):
    base = Path(run_dir)
    try:
        summary = json.loads((base / "summary.json").read_text(encoding="utf-8"))
        manifest = json.loads((base / "run.json").read_text(encoding="utf-8"))
    except FileNotFoundError as exc:
        raise SystemExit("error: %s is not a run directory: %s" % (run_dir, exc))
    except json.JSONDecodeError as exc:
        raise SystemExit("error: invalid JSON in %s: %s" % (run_dir, exc))
    tags_by_id = {}
    for c in manifest.get("cases") or []:
        if isinstance(c, dict) and c.get("id") is not None:
            tags_by_id[c["id"]] = c.get("tags") or {}
    cases = {}
    for c in summary.get("cases") or []:
        if not isinstance(c, dict) or c.get("id") is None:
            continue
        cid = c["id"]
        tags = c.get("tags") if isinstance(c.get("tags"), dict) and c.get("tags") \
            else tags_by_id.get(cid) or {}
        cases[cid] = {
            "value": 1.0 if c.get("passed") else 0.0,
            "family": tags.get("family"),
            "level": tags.get("level"),
            "trace_ref": "%s#%s" % (run_dir, cid),
        }
    return cases


def load_config(config_name):
    rows = [r for r in read_jsonl(CASES_JSONL) if r.get("config_name") == config_name]
    if not rows:
        raise SystemExit("error: no ledger rows for config_name %r in %s" % (config_name, CASES_JSONL))
    cases = {}
    for r in rows:
        cid = r.get("case_id")
        if cid is None:
            continue
        d = cases.setdefault(cid, {"vals": [], "family": None, "level": None,
                                   "trace_ref": r.get("trace_ref")})
        d["vals"].append(1.0 if r.get("passed") else 0.0)
        if d["family"] is None and r.get("family"):
            d["family"] = r.get("family")
        if d["level"] is None and r.get("level"):
            d["level"] = r.get("level")
    for d in cases.values():
        d["value"] = sum(d["vals"]) / len(d["vals"])
    return cases


def load_side(arg):
    if Path(arg).is_dir():
        return load_run_dir(arg)
    return load_config(arg)


def bootstrap_ci(families, a_vals, b_vals, n_boot=N_BOOT, seed=BOOT_SEED):
    """Percentile CI of mean(a)-mean(b) resampling whole family groups."""
    rng = random.Random(seed)
    diffs = []
    n_fam = len(families)
    for _ in range(n_boot):
        a_sum = b_sum = n = 0
        for _ in range(n_fam):
            fam = families[rng.randrange(n_fam)]
            for i in fam:
                a_sum += a_vals[i]
                b_sum += b_vals[i]
                n += 1
        if n:
            diffs.append((a_sum - b_sum) / n)
    if not diffs:
        return None, None, 0.0
    diffs.sort()
    lo = diffs[int(0.025 * (len(diffs) - 1))]
    hi = diffs[int(round(0.975 * (len(diffs) - 1)))]
    point = sum(a_vals) / len(a_vals) - sum(b_vals) / len(b_vals)
    return lo, hi, point


def main(argv=None):
    ap = argparse.ArgumentParser(
        description="Compare two runs (directories) or two config names (ledger) case by case.")
    ap.add_argument("a", help="run dir A or config name A")
    ap.add_argument("b", help="run dir B or config name B")
    args = ap.parse_args(argv)

    a_cases = load_side(args.a)
    b_cases = load_side(args.b)
    common = sorted(set(a_cases) & set(b_cases))
    only_a = sorted(set(a_cases) - set(b_cases))
    only_b = sorted(set(b_cases) - set(a_cases))

    a_pass = [cid for cid in common if a_cases[cid]["value"] >= PASS_THRESHOLD]
    b_pass = [cid for cid in common if b_cases[cid]["value"] >= PASS_THRESHOLD]
    flips_a_pass_b_fail = [cid for cid in common
                           if a_cases[cid]["value"] >= PASS_THRESHOLD
                           and b_cases[cid]["value"] < PASS_THRESHOLD]
    flips_a_fail_b_pass = [cid for cid in common
                           if a_cases[cid]["value"] < PASS_THRESHOLD
                           and b_cases[cid]["value"] >= PASS_THRESHOLD]

    a_vals = [a_cases[cid]["value"] for cid in common]
    b_vals = [b_cases[cid]["value"] for cid in common]

    families = {}
    for i, cid in enumerate(common):
        fam = a_cases[cid]["family"] or b_cases[cid]["family"] or "__solo__:%s" % cid
        families.setdefault(fam, []).append(i)
    family_lists = list(families.values())
    lo, hi, point = bootstrap_ci(family_lists, a_vals, b_vals)

    def flip_entries(cids):
        return [{"case": cid,
                 "a_trace": a_cases[cid]["trace_ref"],
                 "b_trace": b_cases[cid]["trace_ref"]} for cid in cids]

    report = {
        "a": args.a,
        "b": args.b,
        "common_cases": len(common),
        "pass_rate": {
            "a": round(sum(a_vals) / len(a_vals), 4) if a_vals else None,
            "b": round(sum(b_vals) / len(b_vals), 4) if b_vals else None,
            "diff_a_minus_b": round(point, 4) if a_vals and b_vals else None,
        },
        "bootstrap": {
            "unit": "family",
            "replicates": N_BOOT,
            "seed": BOOT_SEED,
            "families": len(family_lists),
            "ci95_a_minus_b": [None if lo is None else round(lo, 4),
                               None if hi is None else round(hi, 4)],
        },
        "flips": {
            "a_pass_b_fail": flip_entries(flips_a_pass_b_fail),
            "a_fail_b_pass": flip_entries(flips_a_fail_b_pass),
        },
        "only_in_a": only_a,
        "only_in_b": only_b,
    }
    print(json.dumps(report, ensure_ascii=False, indent=2))

    print()
    print("== %s vs %s ==" % (args.a, args.b))
    print("common cases: %d (only in A: %d, only in B: %d)"
          % (len(common), len(only_a), len(only_b)))
    if a_vals and b_vals:
        print("pass rate: A %.1f%%  B %.1f%%  diff (A-B) %+.1fpp"
              % (100 * sum(a_vals) / len(a_vals), 100 * sum(b_vals) / len(b_vals), 100 * point))
        print("bootstrap 95%% CI of diff (family resample, n=%d): [%+.1fpp, %+.1fpp]"
              % (N_BOOT, 100 * lo, 100 * hi))
    print("flips: A pass / B fail: %s" % (flips_a_pass_b_fail or "none"))
    print("flips: A fail / B pass: %s" % (flips_a_fail_b_pass or "none"))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: internal/lab/runs/compare.go
package runs

import (
	"fmt"
	"path/filepath"
	"sort"
	"strings"

	"github.com/no22/RWKV-Agent/internal/lab"
)

// Compare two workbank runs or configs case by case.
//
// For each side a per-case pass value is built: 1.0/0.0 for run directories,
// the mean over k replicas for config names. Cases are aligned by case_id and
// the report carries the flip list plus a bootstrap 95% CI of the pass-rate
// difference (A - B), resampling whole families because same-family variants
// are highly correlated.

const (
	nBoot         = 2000
	bootSeed      = 0
	passThreshold = 0.5
)

type compareCase struct {
	value    float64
	count    int
	family   string
	level    string
	traceRef string
}

// CompareArgs are the `run compare` flags.
type CompareArgs struct {
	A string
	B string
}

// RunCompare is the `run compare` command.
func RunCompare(args CompareArgs) int {
	aCases, err := loadSide(args.A)
	if err != nil {
		fmt.Fprintf(stderr, "error: %s\n", err)
		return 1
	}
	bCases, err := loadSide(args.B)
	if err != nil {
		fmt.Fprintf(stderr, "error: %s\n", err)
		return 1
	}

	var common, onlyA, onlyB []string
	for cid := range aCases {
		if _, ok := bCases[cid]; ok {
			common = append(common, cid)
		} else {
			onlyA = append(onlyA, cid)
		}
	}
	for cid := range bCases {
		if _, ok := aCases[cid]; !ok {
			onlyB = append(onlyB, cid)
		}
	}
	sort.Strings(common)
	sort.Strings(onlyA)
	sort.Strings(onlyB)

	var flipsAPassBFail, flipsAFailBPass []string
	for _, cid := range common {
		a, b := aCases[cid].value, bCases[cid].value
		if a >= passThreshold && b < passThreshold {
			flipsAPassBFail = append(flipsAPassBFail, cid)
		}
		if a < passThreshold && b >= passThreshold {
			flipsAFailBPass = append(flipsAFailBPass, cid)
		}
	}

	aVals := make([]float64, 0, len(common))
	bVals := make([]float64, 0, len(common))
	for _, cid := range common {
		aVals = append(aVals, aCases[cid].value)
		bVals = append(bVals, bCases[cid].value)
	}

	// Families are grouped in the order the sorted common list introduces them,
	// because the bootstrap indexes into this list.
	familyIndex := map[string]int{}
	var families [][]int
	for i, cid := range common {
		fam := aCases[cid].family
		if fam == "" {
			fam = bCases[cid].family
		}
		if fam == "" {
			fam = "__solo__:" + cid
		}
		idx, ok := familyIndex[fam]
		if !ok {
			families = append(families, nil)
			idx = len(families) - 1
			familyIndex[fam] = idx
		}
		families[idx] = append(families[idx], i)
	}

	lo, hi, point, hasCI := bootstrapCI(families, aVals, bVals)

	flipEntries := func(cids []string) []any {
		out := make([]any, 0, len(cids))
		for _, cid := range cids {
			entry := lab.NewOrderedMap()
			entry.Set("case", cid)
			entry.Set("a_trace", aCases[cid].traceRef)
			entry.Set("b_trace", bCases[cid].traceRef)
			out = append(out, entry)
		}
		return out
	}

	passRate := lab.NewOrderedMap()
	if len(aVals) > 0 {
		passRate.Set("a", lab.PyFloat(lab.RoundHalfEven(sum(aVals)/float64(len(aVals)), 4)))
	} else {
		passRate.Set("a", nil)
	}
	if len(bVals) > 0 {
		passRate.Set("b", lab.PyFloat(lab.RoundHalfEven(sum(bVals)/float64(len(bVals)), 4)))
	} else {
		passRate.Set("b", nil)
	}
	if len(aVals) > 0 && len(bVals) > 0 {
		passRate.Set("diff_a_minus_b", lab.PyFloat(lab.RoundHalfEven(point, 4)))
	} else {
		passRate.Set("diff_a_minus_b", nil)
	}

	ci := []any{nil, nil}
	if hasCI {
		ci = []any{lab.PyFloat(lab.RoundHalfEven(lo, 4)), lab.PyFloat(lab.RoundHalfEven(hi, 4))}
	}
	bootstrap := lab.NewOrderedMap()
	bootstrap.Set("unit", "family")
	bootstrap.Set("replicates", nBoot)
	bootstrap.Set("seed", bootSeed)
	bootstrap.Set("families", len(families))
	bootstrap.Set("ci95_a_minus_b", ci)

	flips := lab.NewOrderedMap()
	flips.Set("a_pass_b_fail", flipEntries(flipsAPassBFail))
	flips.Set("a_fail_b_pass", flipEntries(flipsAFailBPass))

	report := lab.NewOrderedMap()
	report.Set("a", args.A)
	report.Set("b", args.B)
	report.Set("common_cases", len(common))
	report.Set("pass_rate", passRate)
	report.Set("bootstrap", bootstrap)
	report.Set("flips", flips)
	report.Set("only_in_a", toAnySlice(onlyA))
	report.Set("only_in_b", toAnySlice(onlyB))

	data, err := lab.EncodeOrderedJSON(report, lab.EncodeOptions{Indent: 2})
	if err != nil {
		fmt.Fprintln(stderr, err)
		return 1
	}
	fmt.Println(string(data))

	fmt.Println()
	fmt.Printf("== %s vs %s ==\n", args.A, args.B)
	fmt.Printf("common cases: %d (only in A: %d, only in B: %d)\n", len(common), len(onlyA), len(onlyB))
	if len(aVals) > 0 && len(bVals) > 0 {
		fmt.Printf("pass rate: A %.1f%%  B %.1f%%  diff (A-B) %+.1fpp\n",
			100*sum(aVals)/float64(len(aVals)), 100*sum(bVals)/float64(len(bVals)), 100*point)
		fmt.Printf("bootstrap 95%% CI of diff (family resample, n=%d): [%+.1fpp, %+.1fpp]\n",
			nBoot, 100*lo, 100*hi)
	}
	fmt.Printf("flips: A pass / B fail: %s\n", listOrNone(flipsAPassBFail))
	fmt.Printf("flips: A fail / B pass: %s\n", listOrNone(flipsAFailBPass))
	return 0
}

// bootstrapCI is a percentile CI of mean(a)-mean(b) resampling whole family
// groups. The stream is Python's random.Random(0), because these intervals are
// written into reports (P7).
func bootstrapCI(families [][]int, aVals, bVals []float64) (lo, hi, point float64, ok bool) {
	rng := lab.NewPyRandom(bootSeed)
	var diffs []float64
	nFam := len(families)
	for i := 0; i < nBoot; i++ {
		aSum, bSum, n := 0.0, 0.0, 0
		for j := 0; j < nFam; j++ {
			fam := families[rng.RandRange(nFam)]
			for _, idx := range fam {
				aSum += aVals[idx]
				bSum += bVals[idx]
				n++
			}
		}
		if n > 0 {
			diffs = append(diffs, (aSum-bSum)/float64(n))
		}
	}
	if len(diffs) == 0 {
		return 0, 0, 0, false
	}
	sort.Float64s(diffs)
	lo = diffs[int(0.025*float64(len(diffs)-1))]
	hi = diffs[int(lab.RoundHalfEven(0.975*float64(len(diffs)-1), 0))]
	point = sum(aVals)/float64(len(aVals)) - sum(bVals)/float64(len(bVals))
	return lo, hi, point, true
}

func sum(values []float64) float64 {
	total := 0.0
	for _, v := range values {
		total += v
	}
	return total
}

func loadSide(arg string) (map[string]compareCase, error) {
	if info, err := statPath(arg); err == nil && info {
		return loadRunDir(arg)
	}
	return loadConfig(arg)
}

func loadRunDir(runDir string) (map[string]compareCase, error) {
	summary, err := LoadJSONFile(filepath.Join(runDir, "summary.json"), false)
	if err != nil || summary == nil {
		return nil, fmt.Errorf("%s is not a run directory", runDir)
	}
	manifest, err := LoadJSONFile(filepath.Join(runDir, "run.json"), false)
	if err != nil || manifest == nil {
		return nil, fmt.Errorf("%s is not a run directory", runDir)
	}
	tagsByID := map[string]map[string]any{}
	for _, c := range mapSlice(manifest["cases"]) {
		if id, ok := c["id"].(string); ok {
			tagsByID[id] = mapOf(c, "tags")
		}
	}
	cases := map[string]compareCase{}
	for _, c := range mapSlice(summary["cases"]) {
		cid, ok := c["id"].(string)
		if !ok {
			continue
		}
		tags := mapOf(c, "tags")
		if len(tags) == 0 {
			tags = tagsByID[cid]
		}
		if tags == nil {
			tags = map[string]any{}
		}
		value := 0.0
		if passed, _ := c["passed"].(bool); passed {
			value = 1.0
		}
		family, _ := tags["family"].(string)
		cases[cid] = compareCase{
			value:    value,
			count:    1,
			family:   family,
			level:    stringOf(tags, "level"),
			traceRef: fmt.Sprintf("%s#%s", runDir, cid),
		}
	}
	return cases, nil
}

func loadConfig(configName string) (map[string]compareCase, error) {
	cases := map[string]compareCase{}
	var order []string
	for _, r := range ReadJSONL(DefaultLedgerCases()) {
		if stringOf(r, "config_name") != configName {
			continue
		}
		cid, ok := r["case_id"].(string)
		if !ok {
			continue
		}
		entry, seen := cases[cid]
		if !seen {
			entry = compareCase{traceRef: stringOf(r, "trace_ref")}
			order = append(order, cid)
		}
		if passed, _ := r["passed"].(bool); passed {
			entry.value += 1.0
		}
		entry.count++
		if entry.family == "" {
			entry.family = stringOf(r, "family")
		}
		if entry.level == "" {
			entry.level = stringOf(r, "level")
		}
		cases[cid] = entry
	}
	if len(order) == 0 {
		return nil, fmt.Errorf("no ledger rows for config_name %s in %s", pyQuote(configName), DefaultLedgerCases())
	}
	for _, cid := range order {
		entry := cases[cid]
		entry.value = entry.value / float64(entry.count)
		cases[cid] = entry
	}
	return cases, nil
}

func listOrNone(items []string) string {
	if len(items) == 0 {
		return "none"
	}
	parts := make([]string, 0, len(items))
	for _, item := range items {
		parts = append(parts, pyQuote(item))
	}
	return "[" + strings.Join(parts, ", ") + "]"
}
